src/tests/neohookean_common.hpp

```diff
@@ -0,0 +1,403 @@
+
+  for ( int i = 0; i < 9; ++i ) {
+    double d2E[9] = { 0.0 };
+    d2E[i] = 1.0;
+    double d2mu = 0.0;
+    double d2lambda = 0.0;
+    __enzyme_fwddiff<void>( (void*)hand_code_deriv, E, d2E, mu, d2mu, lambda, d2lambda, &dW, &d2w );
+    for ( int j = 0; j < 9; ++j ) {
+      d2W_d2E[9 * i + j] = d2w[j];
+    }
+  }
+}
+
+void second_deriv_hand_FD( double* E, double mu, double lambda, double* d2W_d2E, double h = 1e-7 )
+{
+  double E_plus[9] = { 0.0 };
+  double E_minus[9] = { 0.0 };
+  double dw_minus[9] = { 0.0 };
+  double dw_plus[9] = { 0.0 };
+  for ( int i = 0; i < 9; ++i ) {
+    for ( int j = 0; j < 9; ++j ) {
+      E_plus[j] = E[j];
+      E_minus[j] = E[j];
+    }
+    E_plus[i] = E[i] + h;
+    E_minus[i] = E[i] - h;
+    hand_code_deriv( E_plus, mu, lambda, dw_plus );
+    hand_code_deriv( E_minus, mu, lambda, dw_minus );
+    for ( int j = 0; j < 9; ++j ) {
+      d2W_d2E[9 * i + j] = ( dw_plus[j] - dw_minus[j] ) / ( 2 * h );
+    }
+  }
+}
+
+
+void run_fwd_mode(double* E, double mu, double lambda, double* dw_df, int N) {
+    axom::utilities::Timer timer{ false };
+    stress(E, mu, lambda, dw_df);
+    double Dw[9] = {0.0};
+    timer.start();
+    for(int i = 0; i < N; ++i) {
+        stress(E, mu, lambda, dw_df);
+        for(int j = 0; j < 9; ++j) {
+            Dw[j] += dw_df[j];
+        }
+    }
+    timer.stop();
+    std::cout << axom::fmt::format( "Time to calc fwd_diff: {0:f}ms", timer.elapsedTimeInMilliSec() ) << std::endl;
+}
+
+void run_bkwd_mode(double* E, double mu, double lambda, double* dw_dE, int N) {
+    axom::utilities::Timer timer{ false };
+
+    stress_reverse(E, mu, lambda, dw_dE);
+    double Dw[9] = {0.0};
+    
+    timer.start();
+    for(int i = 0; i < N; ++i) {
+        stress_reverse(E, mu, lambda, dw_dE);
+        for(int j = 0; j < 9; ++j) {
+            Dw[j] += dw_dE[j];
+        }
+    }
+    timer.stop();
+    std::cout << axom::fmt::format( "Time to calc backward_diff: {0:f}ms", timer.elapsedTimeInMilliSec() ) << std::endl;
+}
+
+void run_hand_derivative(double* E, double mu, double lambda, double* S, int N) {
+    axom::utilities::Timer timer{ false };
+    hand_code_deriv(E, mu, lambda, S);
+    double Dw[9] = {0.0};
+    timer.start();
+    for(int i = 0; i < N; ++i) {
+        hand_code_deriv(E, mu, lambda, S);
+        for(int j = 0; j < 9; ++j) {
+            Dw[j] += S[j];
+        }
+    }
+    timer.stop();
+    std::cout << axom::fmt::format( "Time to calc hand_diff: {0:f}ms", timer.elapsedTimeInMilliSec() ) << std::endl;
+}
+
+void run_fwd_fwd(double* E, double mu, double lambda, double* S, int N) {
+    axom::utilities::Timer timer{ false };
+    timer.start();
+    for(int i = 0; i < N; ++i) {
+        second_deriv_fwd_fwd(E, mu, lambda, S);
+    }
+    timer.stop();
+    std::cout << axom::fmt::format( "Time to calc fwd_fwd_diff: {0:f}ms", timer.elapsedTimeInMilliSec() ) << std::endl;
+}
+
+void run_rev_fwd(double* E, double mu, double lambda, double* S, int N) {
```

Comment:
Change this `S` too, same with all the other second deriv functions.

Reply:
Let's chat about reducing some of this code duplication with either a template or a std::function parameter.